Fix findfeatures matrix inversion issues and improve FastGeom performance

KrisBecker opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
findfeatures matrix inversion issues are addressed in this PR.

## Description
<!--- Describe your changes in detail -->
Matrix inversion errors have been reported in #4639 that cause findfeatures to abort. This appears to be directly related to problems with the FastGeom procedure. Images that have very little to no overlapping common coverage in their FOVs are typically the source of problems in the ImageSource class.

This PR address the matrix inversion problem with better detection of an invalid matrix. Also, improvements to the FastGeom algorithm were required to detect and minimize the impact on successful use of lists of images, particularly. 

The FastGeom process was implemented in the ImageSource class that made corrections to the algorithm problematic because it lacks parameterization/control of the grid-based geometric mapping correspondence algorithm. Therefore, this algorithm, now referrred to as `Grid`, was reimplemented directly in the FastGeom class. This is an iterative algorithm to that will increase the density of the candidate geometry points in the MATCH image on each successive iteration and run unbounded until every pixel was compared and still failed. 

An additional FastGeom algorithm, called `Radial`, was implemented to compute radial patterns of corresponding geometry patterns with increasing point density on each circle. This algorithm is not iterative but can be configured to produce varying density and radial spacing outward from the center of the image.

As a consequence, these image failures are more quickly and efficiently identified and removed from the matcher list and optionally reported in a user specified file that contains a list of images that cannot be transformed or the matrix inversion fails. Hence, a new parameter to findfeatures was added called TONOGEOM that contains this list.

In addition, a known problem with construction of control points and measures is also addressed in this PR. At times, only the MATCH geometry control measure coordinate was set in the newly created control point. However, the corresponding geometry control measure in the FROM/FROMLIST images was never checked/validated. This caused downstream processing problems, particularly in jigsaw.

Lots of debugging information has been added to help in better understanding and track performance of FastGeom and creation of the output control network.

Here is the summary of changes contained in this PR:

* Fix matrix inversion error on empty matrix. (Fixes #4639)
* Identify images that fail FastGeom transform and exclude from matching.
* Add TONOGEOM parameter to write failed file loads or FastGeom error file list if the transform cannot be determined
* Improve FastGeom transform algorithm using new Radial point mapping scheme
* Add more debugging output to help diagnose problem images/procedures
* findfeatures modified to add Grid algorithm to FastGeom class
* Improved Grid algorithm by computing proper starting iteration to statisfy fastgeompoints request
* Added additional parameterization of Grid algorithm
* Reorganized mapping process to consolidate Radial and Grid algorithms in FastGeom
* Fixed fetching of target name in ImageSource that supports use of projected images/mosaics

## Related Issue
<!--- This project only accepts pull requests related to open issues (GitIssues) -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->
Fixes #4639.
Fixes #4771.

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
findfeatures would abort on single images when applying FastGeom processing to computed a warp transform to better spatially compare the MATCH image to FROM/FROMLIST images. This allows use of OpenCV feature/detector algorithms that are not rotation and/or scale invariant.

Problems with a single image would abort findfeatures when they could simply be excluded from the matching process and continue processing images that are successfully processed.

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

These changes have been tested in for several months processing NEAR/MSI Eros and Rosetta/OSIRIS-NAC 67P images. Modifications to findfeatures affect only code used in the application and does not impact any other system classes or functionality.

## Screenshots (if appropriate):

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Documentation change (update to the documentation; no code change)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
<!--- - [ ] My code follows the code style of this project. -->
- [x] I have read and agree to abide by the [Code of Conduct](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/Code-Of-Conduct.md)
- [x] I have read the [**CONTRIBUTING**](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/CONTRIBUTING.md) document.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] I have added myself to the [.zenodo.json](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/.zenodo.json) document.
- [ ] I have added any user impacting changes to the [CHANGELOG.md](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/CHANGELOG.md) document.

## Licensing
This project is mostly composed of free and unencumbered software released into the public domain, and we are unlikely to accept contributions that are not also released into the public domain. Somewhere near the top of each file should have these words:

> This work is free and unencumbered software released into the public domain. In jurisdictions that recognize copyright laws, the author or authors of this software dedicate any and all copyright interest in the software to the public domain.

- [x] I dedicate any and all copyright interest in this software to the public domain. I make this dedication for the benefit of the public at large and to the detriment of my heirs and successors. I intend this dedication to be an overt act of relinquishment in perpetuity of all present and future rights to this software under copyright law.

KrisBecker committed:
findfeatures bug fixes and improvements

* Fix matrix inversion error on empty matrix. (Fixes #4639)
* Identify images that fail FastGeom transform and exclude from matching.
* Add TONOGEOM parameter to write failed file loads or FastGeom error file list if the transform cannot be determined
* Improve FastGeom transform algorithm using new radial point mapping scheme
* Add more debugging output to help diagnose problem images/procedures

KrisBecker committed:
findfeatures - add Grid algorithm to FastGeom

* findfeatures mod to add Grid algorithm to FastGeom class
* Improved Grid algorithm by computing proper starting iteration to statisfy fastgeompoints request
* Added additional parameterization of Grid algorithm
* Reorganized mapping process to consolidate Radial and Grid algorithms in FastGeom

KrisBecker committed:
Merge branch 'USGS-Astrogeology:dev' into findfeatures_matrix_fix_mods

jessemapel commented on isis/src/control/apps/findfeatures/FastGeom.h:
```diff
@@ -75,6 +94,21 @@ class FastGeom {
+          return (true);
+        }
+      }
+      return (false);
```
This can be reduced down to

```
return (point.x >= xmin ) && (point.x < xmax ) && ( point.y >= ymin ) && (point.y < ymax )
```

I assume the compiler will optimize this away, but it's much more readable than a double nested if that just returns a boolean.

KrisBecker replied:
I completely removed this method in favor of OpenCV Rect::contains. It has the same comparison.

jessemapel commented on isis/src/control/apps/findfeatures/GenericTransform.cpp:
```diff
@@ -120,6 +122,35 @@ cv::Point2f GenericTransform::inverse(const cv::Point2f &point) const {
+                                         const bool verify) {
+  cv::Mat inverse;
+  double result = cv::invert(matrix, inverse);
+  if ( verify == true ) {
```
Just do `if (verify)` here

KrisBecker replied:
Done.

jessemapel commented on isis/src/control/apps/findfeatures/GenericTransform.cpp:
```diff
@@ -120,6 +122,35 @@ cv::Point2f GenericTransform::inverse(const cv::Point2f &point) const {
+  cv::Mat inverse;
+  double result = cv::invert(matrix, inverse);
+  if ( verify == true ) {
+    if ( qFuzzyCompare(result+1.0, 0.0+1.0) ) {
```
Why do a fuzzy comparison? The OpenCV docs say that it will return exactly 0 if the matrix is singular.

https://docs.opencv.org/4.5.3/d2/de8/group__core__array.html#gad278044679d4ecf20f7622cc151aaaa2

KrisBecker replied:
I actually looked at the [code](https://github.com/opencv/opencv/blob/8b4fa2605e1155bbef0d906bb1f272ec06d7796e/modules/core/src/lapack.cpp#L843-L1047) and it returns a boolean. I assumed it would be the determinant, hence the fuzzy compare. I changed it as you suggested.

jessemapel commented on isis/src/control/apps/findfeatures/GenericTransform.cpp:
```diff
@@ -120,6 +122,35 @@ cv::Point2f GenericTransform::inverse(const cv::Point2f &point) const {
+  double result = cv::invert(matrix, inverse);
+  if ( verify == true ) {
+    if ( qFuzzyCompare(result+1.0, 0.0+1.0) ) {
+      QString msg = "cv::Mat inverse matrix is not invertable";
```
Looks like there's an extra inverse and invertable should be invertible. Error message should be something like "Transformation matrix is not invertible"

KrisBecker replied:
Reworded the error message to actually make sense.

jessemapel commented on isis/src/control/apps/findfeatures/FastGeom.cpp:
```diff
@@ -62,24 +67,398 @@ FastGeom::~FastGeom() { }
+ImageTransform *FastGeom::compute(MatchImage &query, MatchImage &train,
+                                  QLogger logger) const {
+
+  // Shortcut...
```
What does this comment mean?

KrisBecker replied:
Its a shortcut (reference) to the actual stream operator. I expanded the comment to explain this.

jessemapel commented on isis/src/control/apps/findfeatures/FastGeom.cpp:
```diff
@@ -62,24 +67,398 @@ FastGeom::~FastGeom() { }
+  double q_max_samp = q_samps + 0.4999 + fg_q_sample_tolerance;
+  double q_min_line = 0.5 - fg_q_line_tolerance;
+  double q_max_line = q_lines + 0.4999 + fg_q_line_tolerance;
+  FGFov  q_fov(q_min_samp, q_min_line, (q_max_samp-q_min_samp+1.0), (q_max_line-q_min_line+1.0) );
```
Why add 1 to the FoV sizes?

jessemapel replied:
This variable is entirely unused

KrisBecker replied:
Thats a bug, which is fixed. I retained the variable in favor of minimizing the number of variables and use of the OpenCV Rect type to test for points in the image FOV.

jessemapel commented on isis/src/control/apps/findfeatures/FastGeom.cpp:
```diff
@@ -62,24 +67,398 @@ FastGeom::~FastGeom() { }
+  logit <<   "  FastGeomLineTolerance:   " << fg_line_tolerance << "\n\n";
+
+  // Run the requested algorithm
+  if ( true == radial_method) {
```
no need to compare against true, just `if (radial_method)`

KrisBecker replied:
Done.

jessemapel commented on isis/src/control/apps/findfeatures/FastGeom.cpp:
```diff
@@ -62,24 +67,398 @@ FastGeom::~FastGeom() { }
+
+    // Compute maximum radial track at the center of the image to
+    // the corner and scaling parameters
+    double fg_max_radius    = std::sqrt( (q_samps * q_samps) + ( q_lines * q_lines) );
```
This is the maximum diameter, not the maximum radius. Should be divided by 2 to get the distance from the center of the image to a corner.

KrisBecker replied:
Thats a bug. Fixed.

jessemapel commented on isis/src/control/apps/findfeatures/FastGeom.cpp:
```diff
@@ -62,24 +67,398 @@ FastGeom::~FastGeom() { }
+
+    // No check for explict specs
+    // Allow user explicit radius delta increments - minimum is the length of the
+    // diagonal of 1 pixel ( 2 * sqrt(1^2 + 1^2))
```
This the diagonal length of 2 pixels. One pixel is just sqrt(2) diagonally.

KrisBecker replied:
The comment is incorrect. I want to avoid testing the same pixel (corners). So I set it as indicated in the code and updated the comment.

jessemapel commented on isis/src/control/apps/findfeatures/FastGeom.cpp:
```diff
@@ -62,24 +67,398 @@ FastGeom::~FastGeom() { }
+        // scaled by the radial pixel distance from the center of
+        // the image at (c_x, c_y).
+        double q_x = ( ((v_x0 * v_cos) + (v_y0 * -v_sin)) * fg_radius ) + c_x;
+        double q_y = ( ((v_x0 * v_sin) + (v_y0 *  v_cos)) * fg_radius ) + c_y;
```
You don't need a reference vector here as it cannot be parameterized, and it's already implicit in the sin and cos functions. These lines can just be

```
double q_x = c_x + (v_cos * fg_radius)
double q_y = c_y + (v_sin * fg_radius)
```

KrisBecker replied:
Modified as suggested.

jessemapel commented on isis/src/control/apps/findfeatures/FastGeom.cpp:
```diff
@@ -62,24 +67,398 @@ FastGeom::~FastGeom() { }
+        double q_y = ( ((v_x0 * v_sin) + (v_y0 *  v_cos)) * fg_radius ) + c_y;
+
+        // Check to see if the point is in the image FOV
+        FGPoint q_coord(q_x, q_y);
```
Don't bother defining q_x and q_y. Just define the point collectively. This function already has a lot of variable floating around. This will require 239 to use q_coord.x and q_coord.y.

KrisBecker replied:
Ok...

jessemapel commented on isis/src/control/apps/findfeatures/FastGeom.cpp:
```diff
@@ -62,24 +67,398 @@ FastGeom::~FastGeom() { }
+  double t_max_samp = t_samps + 0.4999 + fg_sample_tolerance;
+  double t_min_line = 0.5 - fg_line_tolerance;
+  double t_max_line = t_lines + 0.4999 + fg_line_tolerance;
+  FGFov  t_fov(t_min_samp, t_min_line, (t_max_samp-t_min_samp+1.0), (t_max_line-t_min_line+1.0) );
```
Same as q_fov, this is entirely unused

KrisBecker replied:
Corrected and retained the FGFov variable to test for points in the image.

jessemapel commented on isis/src/control/apps/findfeatures/FastGeom.h:
```diff
@@ -52,19 +53,37 @@ class ImageTransform;
   public:
     typedef GenericTransform::RectArea   RectArea;
+    typedef cv::Point2d                  FGPoint;
+    typedef cv::Rect2d                   FGFov;
```
This typedef is only used for the unused FoV variables and should be removed.

KrisBecker replied:
It is now used to test for points in the FOV so its retained.

jessemapel commented on isis/src/control/apps/findfeatures/FastGeom.cpp:
```diff
@@ -62,24 +67,398 @@ FastGeom::~FastGeom() { }
+              q_infov_points.push_back(q_coord);
+              t_infov_points.push_back(t_coord);
+            }
+            t_inFOV.push_back(inFOV);
```
Put this before the if

KrisBecker replied:
Ok.

jessemapel commented on isis/src/control/apps/findfeatures/FastGeom.cpp:
```diff
@@ -62,24 +67,398 @@ FastGeom::~FastGeom() { }
+    while ( fg_radius <= fg_max_radius ) {
+
+      // Compute the number points / radial
+      double r_points = qMax(1.0, points_per_radius);
```
Why not just define `points_per_radius` as 1 above and then ditching this qMax call?

This could require an explicit check that "FastGeomPointIncrement" is non-negative. As of right now, this will just cause it to always use a single point if "FastGeomPointIncrement" is negative.

KrisBecker replied:
This works as intended. On the first iteration you get exactly one point. On the next iteration you get fg_point_inc points and multiples of the increment on subsequent rings/circles. If fg_point_inc <= 0, you get a line of single points extending from the center of the image to the right. Values 0 < fg_point_inc < 1 might be interesting, but will still work. qMax is an important element of this algorithm.

jessemapel commented on isis/src/control/apps/findfeatures/FastGeom.cpp:
```diff
@@ -62,24 +67,398 @@ FastGeom::~FastGeom() { }
+    // Probably bad, but we will allow it as an extreme.
+    double fg_point_inc = qMax( 0.0, (fg_radius_inc * fg_point_factor) );
+    if ( m_parameters.exists("FastGeomPointIncrement") ) {
+      fg_point_inc = toDouble(m_parameters.get("FastGeomPointIncrement"));
```
Why are both the factor and increments user settable? Is there a use case where you want to set the factor over the increment?

KrisBecker replied:
Allowing both to be user settable provides for a lot of flexibility. This is useful for irregular bodies and approach observations to provide point density.

jessemapel commented on isis/src/control/apps/findfeatures/FastGeom.cpp:
```diff
@@ -62,24 +67,398 @@ FastGeom::~FastGeom() { }
+    mapper = train.source().getGeometryMapping(query.source(),
+                                               m_fastpts,
+                                               m_tolerance);
+#endif
```
Remove this as we don't set compiler flags. If users would reasonably want to use the old algorithm then it should be an option like grid vs ring.

KrisBecker replied:
Removed. This was for reference only.

jessemapel commented on isis/src/control/apps/findfeatures/FastGeom.cpp:
```diff
@@ -62,24 +67,398 @@ FastGeom::~FastGeom() { }
+
+      // Update radial distance from image center and points/circle
+      fg_radius         += fg_radius_inc;
+      points_per_radius += fg_point_inc;
```
What is the goal for the point distribution on the rings? Should points be equally spaced on the ring (equal arc length between then), equally spaced spatially (equal euclidean distance between them), something less regular?

With the pattern as implemented, you'll get a big variety of spatial point densities.

KrisBecker replied:
Thats the point of this algorithm - to provide density when needed. Note the radial method does not loop and reconfigure the density as the grid algorithm does. You got one shot to get enough points.

jessemapel commented on isis/src/control/apps/findfeatures/FastGeom.cpp:
```diff
@@ -62,24 +67,398 @@ FastGeom::~FastGeom() { }
+    while ( fg_radius <= fg_max_radius ) {
+
+      // Compute the number points / radial
+      double r_points = qMax(1.0, points_per_radius);
```
Would you want to floor this and make r_points an int so that the points are equally spread around the ring? With this implementation, you'll get "fractional" points in the lower right corner of the image.

KrisBecker replied:
That is an implementation choice. This is something I can test but I think it would produce strange patterns due to other floating point elements of the algorithm.

jessemapel commented on isis/src/control/apps/findfeatures/FastGeom.cpp:
```diff
@@ -62,24 +67,398 @@ FastGeom::~FastGeom() { }
+    logit << "--> Using Grid Algorithm train-to-query mapping <--\n";
+
+     // Compute increment
+    int   fg_minpts = qMax(m_fastpts, 16);
```
Odd spacing here

KrisBecker replied:
Fixed/reworked.

jessemapel commented on isis/src/control/apps/findfeatures/FastGeom.cpp:
```diff
@@ -62,24 +67,398 @@ FastGeom::~FastGeom() { }
+    int iteration = fg_grid_start_iter;
+    int currinc(0);
+    double v_q_nsamps = q_max_samp - q_min_samp + 1.0;
+    double v_q_nlines = q_max_line - q_min_line + 1.0;
```
Why +1 on these? q_min/max_samp/line do not appear to be inclusive based on how they are defined.

KrisBecker replied:
Fixed.

jessemapel commented on isis/src/control/apps/findfeatures/FastGeom.cpp:
```diff
@@ -62,24 +67,398 @@ FastGeom::~FastGeom() { }
+      double sSpacing = qMax( 1.0, v_q_nsamps/(currinc*1.0) );
+      double lSpacing = qMax( 1.0, v_q_nlines/(currinc*1.0) );
+
+      if ( qMax( sSpacing, lSpacing ) <= 1.0 ) done = true; // Last possible loop
```
The max here seems needlessly complex, just check if they are both 1.0

jessemapel commented on isis/src/control/apps/findfeatures/FastGeom.cpp:
```diff
@@ -62,24 +67,398 @@ FastGeom::~FastGeom() { }
+
+     // Compute increment
+    int   fg_minpts = qMax(m_fastpts, 16);
+    int increment = (int) (std::sqrt(std::max(24.0, (double) (fg_minpts - 1))) + 1.0);
```
Use a ceil call instead of subtracting and adding 1 before casting to an int. This will make it much easier to understand what this is trying to do. So, it would look like:

```
int increment = ceil(std::sqrt(std::max(24.0, fg_minpts)));
```

KrisBecker replied:
Ok.

jessemapel commented on isis/src/control/apps/findfeatures/FastGeom.cpp:
```diff
@@ -62,24 +67,398 @@ FastGeom::~FastGeom() { }
+          double q_x = (sSpacing / 2.0 + sSpacing * s + 0.5) + q_min_samp;
+
+          // Check to see if the point is in the image FOV
+          FGPoint q_coord(q_x, q_y);
```
Same as the ring algorithm, you may as well just define the point directly.

KrisBecker replied:
Done.

jessemapel commented on isis/src/control/apps/findfeatures/FastGeom.cpp:
```diff
@@ -98,43 +477,98 @@ ImageTransform *FastGeom::compute(MatchImage &query, MatchImage &train)  {
   }
 
   return ( fastg.take() );
 }
```
This is a really big function doing a lot of things and has a lot of inputs. It would be good to split out some of the algorithms so that they can be properly unit tested and we can ensure that we're properly checking and handling all of the parameterization.

KrisBecker replied:
I made separate methods for the grid, radial and dump portions of the compute() function and made them public scope so they can be directly called.

jessemapel commented on isis/src/control/apps/findfeatures/findfeatures.cpp:
```diff
@@ -88,12 +90,69 @@ namespace Isis{
+   * @param matcher    The source reponsible for managing the matcher process
+   * @param trainfile  File name to load
+   * @param fastgeom   Apply FastGeom if provded. Errors will return false with
+   *                      no adding to the matcher
```
Looks like a typo, should be "return false without adding to the matcher"

KrisBecker replied:
Fixed.

jessemapel commented on isis/src/control/apps/findfeatures/findfeatures.cpp:
```diff
@@ -88,12 +90,69 @@ namespace Isis{
+      isGood = true;
+    }
+    catch ( IException &ie ) {
+      isGood = false;
```
Why not just return false here and return true at the end?

KrisBecker replied:
Ok

jessemapel commented on isis/src/control/apps/findfeatures/findfeatures.cpp:
```diff
@@ -327,15 +433,22 @@ namespace Isis{
       cnet.SetTarget( target );
       ID pointId( ui.GetString("POINTID"), ui.GetInteger("POINTINDEX") );
-      matcher.setDebugOff();
+      // matcher.setDebugOff();  TURN ON DEBUG OUTPUT FOR NETWORK GENERATION!
```
Why is this here? If it needs to be on just delete this line

KrisBecker replied:
This indicates a behavioral change. It was explicitly turned off in the commented code and I left it in the initial PR to indicate that change. I have since removed it.

jessemapel commented on isis/src/control/apps/findfeatures/findfeatures.cpp:
```diff
@@ -327,15 +433,22 @@ namespace Isis{
         QString mess = "No control points found!!";
         logger->dbugout() << mess << "\n";
+
+        // Get total elapded time
```
typo, elapsed time

KrisBecker replied:
Fixed.

jessemapel reviewed (commented):
Overall looks good. I have a few questions about constant choices and what the desired matching patterns are. Most of the comments are just readability and clean-up.

The point distribution and matching algorithm is quite complex, it would be really helpful to separate it out so that it can be unit tested more easily.

KrisBecker commented:
I am currently testing these modifications and will be ready to push them to the branch tomorrow unless you have additional comments/concerns.

KrisBecker committed:
findfeatures modifications after Astro code review

findfeatures modifications:

* Refactored FastGeom separating large sections of the compute() methods
* Redesigned the the radial algorithm for easier user configuration
* Added sanity checks avoid bad geometric correspondences
* Added GLOBALS parameters for easy configuration of FastGeom algorithms at runtime

KrisBecker committed:
Merge branch 'findfeatures_matrix_fix_mods' of https://github.com/KrisBecker/ISIS3 into findfeatures_matrix_fix_mods

KrisBecker commented:
Thanks @jessemapel for your very helpful feedback. Your comments compelled a deeper dive in testing and evaluation of the modifications. This identified additional issues and errors that produced a rather large refactor.

I pushed those updates into my branch that is ready for review. Clearly this app begs fo an overarching guide to help make it easier to use effectively and productively. Its something on my todo list for sure.

Please let me know if you have any other concerns.

jessemapel commented:
Code looks good. For testing, I think the best way to handle this is to have simple tests for findfeatures that ensure it's calling the correct algorithm between grid and ring. Then, the grid and ring algorithms can be unittested to ensure that all of the parameters are being passed in and used correctly.

jlaura commented:
@KrisBecker @jessemapel What is the status on this? Looks like it is waiting on tests?

jessemapel commented:
Correct this just needs tests and it's ready to go

rfergason commented:
This pull request is still active and should remain open.  Thanks!

KrisBecker committed:
Merge remote-tracking branch 'origin/dev' into findfeatures_matrix_fix_mods

KrisBecker committed:
Merge branch 'USGS-Astrogeology:dev' into findfeatures_matrix_fix_mods

KrisBecker committed:
Fixed existing findfeatures test - error text

The ThreeImageNetwork.FunctionalTestFindfeaturesErrorNoInput expects an error to be thrown with spectific text. The text in findfeatures.cpp was changed in this PR. It fixes the existing findfeatures test suite.

KrisBecker committed:
Merge branch 'findfeatures_matrix_fix_mods' of https://github.com/KrisBecker/ISIS3 into findfeatures_matrix_fix_mods

KrisBecker committed:
FastGeom.cpp updated to better accomodate testing

KrisBecker committed:
Add findfeatures Radial/Grid config files for test

KrisBecker committed:
Updated findfeatures test for Grid/Radial algos

KrisBecker committed:
Fixed cnetwinnow test that created misplaced files

The cnetwinnow test needs tempDir().path() prepended to the “file_prefix” parameter to properly place the output files in the teardown directory.

KrisBecker commented:
Ok, I think these modifications address the remaining issues. Could someone please review this.

Thanks for your patience.

KrisBecker committed:
Small adjustment to new findfeatures tests

KrisBecker commented:
Pinging this PR for review/merge. 

Thanks...

acpaquette commented on isis/src/control/apps/findfeatures/findfeatures.xml:
```diff
@@ -1543,6 +1589,33 @@ End
           <filter>*.conf</filter>
       </parameter>
 
+      <parameter name= "GLOBALS">
+            <type>string</type>
+            <brief>String containing global parameters and values</brief>
+            <description>
+                <p>
+                    This string can contain additional parameters that will
+                    set or reset global parameters provded by other mechanisms.
+                    This program option is primarily useful for making small
+                    scale adjustments to algorithm parameters in a convenient
+                    and efficient manner.
+                </p>
+                <p>
+                    For example, using this parameter is the most straightforward
+                    way to select the FASTGEOM algorithm. To select the FastGeom
+                    Grid algorithm and constrain the number of interations to 10,
+                    use GLOBALS=FastGeomAlgorithm:Grid@FastGeomGridStopIteration:10".
+                </p>
+                <p>
+                    There is very little robust error detection or any confirmation
+                    that the parameter names are valid. Any misspelled parameters
+                    are not detected and ill-formed parameter strings may not
+                    result in errors, but passthrough and mangle other parameter
+                    construction.
+                </p>
+            </description>
+          <internalDefault>None</internalDefault>
+      </parameter>
```
This seems extremely specific to power users of find features. Without any kind of error checking or other mechanisms it seems like people could really shoot themselves in the foot here.

Plus what variables does this expose? If I dig into the findfeatures code I can find them but they aren't exposed in a way that would even indicate they are present. As a user on the surface I would have no idea how this worked or what it's for.

It seems like this is handled by the `Parameters` input in the application as per the tests, but from the documentation it would seem like the fastgeom parameters should not be part of the `Parameters` file.

I am sure there are a couple ways this could be addressed, but I think adding a `fastgeomparamters` file option if `FastGeom` is enabled would be nice. Either that or this `Global` parameter should only be for `FastGeom` parameters.

More than happy to iterate on a solution here

KrisBecker replied:
I agree. 

This parameter is very useful for users mainly because it saves them from having to provide parameters in a file - this is very inconvenient when running in batch - as there is no other way to alter the parameters otherwise.

After a scan of the code and documentation, I see there really is no description how to use FASTGEOM. Note that I did [post](https://github.com/DOI-USGS/ISIS3/issues/4771#issuecomment-1029125719) a detailed analysis of these changes and how they impact the process (it will probably go away when [AstroDiscuss](https://astrodiscuss.usgs.gov) goes away. I can add much of this content to the findfeatures documentation, perhaps there is a way I can add this document somewhere in ISIS docs (convert to MD?).

I will also add a section that describes all the parameters in findfeatures that can be modified using this `GLOBAL` parameter. I can also add some templates for both Grid and Radial algorithms in appdata/findfeatures.

I will also add code to dump the full content of the global parameters used in findfeatures during the run in a separate group in the app log (print.prt). I see there is no report of the final parameters used. This will allow users to confirm the desired parameters are being applied.

If its the name of the parameter that is a concern, I chose this name because it does affect the global variable table beyond just the FASTGEOM algorithm. I figured the name `GLOBAL` is a good indicator of that scope. This parameters is required for efficient batch use. If you have a suggest for a more appropriate name/approach, we can certainly discuss that.

Let me know if these suggestions address your concerns or if you have some thoughts.

acpaquette replied:
Maybe something to consider in the future is reworking findfeatures parameter tuning to be a little more consistent. It seems like one could make any change they want from the `GLOBAL` parameter and again end up doing unintended things by overriding variables in a passed in parameter file. For new users it's going to be overwhelming if they have to understand PVL parameter files as well as passing in variables from the `GLOBAL` parameter. At which point they would probably just pick whatever path gives them the least resistance.

Some potential in the future would be having different fields for different parameter sets, getting away from parameter files and moving more towards this key, value text parsing that the `GLOBAL` parameter is doing.

I think adding some context for `FASTGEOM` into the XML for the app somewhere would be ideal, basically context around the FASTGEOM parameters and describing the `GLOBAL` parameter. Seems like a straight forward place for that information to live. 

The other additions also sound useful and would get this closed out!

KrisBecker replied:
The revisions I pushed hopefully resolves these issues/concerns.

Please review and let me know if there is anything remaining.

acpaquette reviewed (commented):
Overall tests look good and test the new FastGeom code. I am a little worried about the new `Global` parameter introduced. See comment below for details.

KrisBecker committed:
Merge branch 'dev' of https://github.com/KrisBecker/ISIS3 into findfeatures_matrix_fix_mods

acpaquette commented:
@KrisBecker Do you have an idea when some of the above discussion will be added to this PR?

KrisBecker commented:
I am currently working on this along with other distractions. I am going to be on vacation later this week into next week so it will be at least two weeks away.

If you are planning a release soon, would it be possible to merge this PR and I submit another one when I have completed this work?

Thanks...

acpaquette commented:
@KrisBecker I think we can make that happen, I will open another issue to get docs on this once we merge this PR

KrisBecker commented:
@acpaquette that would be awesome!

Thank you!

AustinSanders commented:
@KrisBecker I'm in the process of peeling a release at the moment, and I'm in the final stages of merging PRs so that they can be included in 8.0.0 LTS.  I brought this PR up during our standup, and the team is somewhat on the fence about merging this without addressing some of the comments in the most recent review, particularly with respect to missing documentation.

That being said, I recognize that it's important to get this merged.  I can delay the release build until tomorrow morning, but any later than that would make me worried about the release timeline. If there's anything that I can do to help move this along before then, please let me know, and I'll do my best to help so that this can be included in the release-in-progress.

KrisBecker committed:
Merge remote-tracking branch 'origin/dev' into findfeatures_matrix_fix_mods

KrisBecker committed:
Merge branch 'DOI-USGS:dev' into findfeatures_matrix_fix_mods

KrisBecker committed:
Merge branch 'findfeatures_matrix_fix_mods' of https://github.com/KrisBecker/ISIS3 into findfeatures_matrix_fix_mods

KrisBecker committed:
Fixes/improvements to findfeature code

- Improved reporting of parameterizations of findfeatures
- Prevent creation of empty TONOTMATCH file when none are detected

KrisBecker committed:
Significant modifications/improvements to docs

 - Added two new examples demonstrating/documenting the use of FASTGEOM algorithm, parameterization using GLOBALS and how to produce a regional mosaic using findfeatures with batch scripts.
- Reviewed, clarified and improved findfeatures program documentation

KrisBecker committed:
Updated the change log

KrisBecker commented:
Thank you for your patience. 

I have updated the documentation and added some additional examples that addresses concerns. Tables have been added the describes all the parameters that can be applied to modify findfeatures behavior. The examples demonstrate use this feature. 

Please review and let me know if anything remains to get this PR merged.

Thanks for all your suggestions and help with this.

acpaquette commented on isis/appdata/templates/findfeatures/mosaics/00_basic_proc.sh:
```diff
@@ -0,0 +1,188 @@
+#!/bin/bash
```
Just to verify, is there a particular reason to include this chunk of bash scripts under templates? It seems strange to include them here rather that the `isis/scripts` area. Or potentially under some other section for tutorials. ISIS really hasn't shipped these kinds of pipelines in the past. It's something we could do but I am struggling to see a benefit for a new user.

KrisBecker replied:
There are likely several areas where these scripts could be located. I can see why some may argue this is a tutorial rather than a template. Here is my reasoning for including them in findfeatures documentation as an example and why they are in the templates directory for findfeatures:

1. Including this as an example in findfeatures provides context. A lot of effort has been put in this work and this example demonstrates how it is used to produce a small, regional mosaic comprised of nine MESSENGER MDIS images. It explains and demonstrates why many of the design decisions were made (e.g., parameterization, an issue you raised in a previous review).
2. It is a very explicit, small example using a specific dataset. The example explains, with visuals, why the MATCH image was selected and effective application of parameters to fully utilize findfeatures capabilities and features to analyze, evaluate and understand this application. It connects context where this information is provided - in the application documentation.
3. I think putting the scripts in the `isis/scripts` directory may give the false impression they are general/ready enough to use in other circumstance. They are not. For example, the photometric correction is specific to MDIS images. However, they do provide a starting point to gain experience with findfeatures. Users are encouraged to copy, modify to suite and experiment. Putting the scripts in the template directory associates/isolates it to this application as intended.
4. We intend to include, in PRs, at least three broad, comprehensive tutorials with our ISIS APEX work when it is submitted into ISIS. These tutorials will make use of findfeatures and provide a more general, comprehensive discussion/explanation of image processing techniques that can be applied using these features and capabilities.

I hope you have had a chance to see the rendered HTML documentation. I think it is a nice addition to the findfeatures documentation to explain its role in producing mosaics.

This example is provided to help users to better understand findfeatures in a broader context. It provides a baseline configuration (i.e, template) to adjust/modify parameterization that enables quick/efficient evaluations and comparisons of its behavior.

acpaquette replied:
Thanks, this helps me understand this a little better and gives context. While reviewing I was thinking that looking over this rendered may also help see broader scope. I will do that and report back.

If we plan to introduce more comprehensive tutorials would we want to see this info eventually pointed at those or would this remain along side?

Kelvinrr replied:
@KrisBecker Do you have an answer to this? I think it's important to establish the purpose of these script as we need to know to what extent we need to support these scripts. I think there exists a bigger discussion as to how we want to support tutorials. And by merging these script in I believe might set a precedence we do not want to set right now as to how we want to go about supporting tutorial content.  

This also, I believe, further validates the concern of big PRs. It becomes easier to let these slip through the cracks and conflates the conversation with other PR concerns. 

For now I am gonna advocate for their removal from this PR before merging and re-introducing them in a later PR once we agree on a way to manage them.

KrisBecker replied:
@Kelvinrr Simply stated, this is an example. It is an example that shows how findfeatures can be used to produce an image-based network, a concept that is heavily relied upon in the control and production of larger, regional and global mosaics.

Larger scoped tutorials would certainly refer to (the scripts in) this example to help explain what an image-based control network is and describe the basics for generating them using findfeatures. However, the pipelines used to produce the mosaics cited in the example documentation implements findfeatures' creation of these networks differently.

Just like other applications and examples, this example provides a template for users to modified as needed. It also is useful to understand how to use/apply findfeatures and the effects of its parameters.

You certainly raise very good points about discussions/guidelines/policy that need to be had regarding these types of things that currently are not available or are ill-defined. Please open an issue/discussion so that your concerns can be addressed.

Per your request, I will remove the scripts in this example and scrub mention of them from the documentation.

Kelvinrr commented:
I think going forward we need to split these sort of PRs into smaller more manageable chunks. Large PR's almost guarantee this sort of slow progress in the merge process as it's more like merging a large feature branch. For now we still need to focus on wrapping this one up before or during the next support sprint.

KrisBecker commented:
I agree this one got much more difficult over time.

If you think it would help we can arrange for an interactive review. This might help it move along quicker toward merge.

I do see a few typos that I should fix. I may need to restructure the change log entry (its all over the place), so if you have any suggestions there that would be great.

Let me know how I can help get this merged.

Thanks...

acpaquette commented on isis/src/control/apps/findfeatures/findfeatures.xml:
```diff
@@ -3,431 +3,465 @@
+      <i>findfeatures</i> gives users a wide range of options to alter
+      algorithms functionaility. Such a broad range of choices/combinations
+      can be intimidating for users who are unfamiliar with FBM. The following
+      are some sugestions to help make reasonable decision. First,
```
Typo, suggestions instead of sugestions

acpaquette commented on isis/src/control/apps/findfeatures/findfeatures.xml:
```diff
@@ -3,431 +3,465 @@
+      observing conditions during image acquisition. The FASTGEOM parameter
+      helps minimize these challenges by using a priori ephemeris SPICE data
+      to create and apply a perspective matrix warp projection using the
+      query and training geometry. Combining the FASTGEOM option with
+      algorithms that are designed for speed (the FAST descriptor and BRIEF
+      extractor are reasonable options) will quickly produce a high quality
+      control network.
```
This seems like a very general and difficult to prove statement. To what standard is the control network high quality?

Maybe changing some wording to say, "...will quickly produce a control network. This process typically creates robust control networks but should be verified by the user"

KrisBecker replied:
Fair enough.

There are at least two ways to assess network quality - numerically and visually.

jigsaw produces a lot of bundleout data that I have used for numerical assessment. I look for a standard residual of no more than 1 pixel in measures/points. I also look for image RMS less that a pixel. There is more analysis that can be done, but these are the fundamentals.

Visual inspection of the controlled mosaic is the ultimate indication of quality.

That particular comment is based upon personal experience. 

I do agree its misleading to imply the output result of findfeatures is directly a high quality network. A quality network may exist in that network, but application of other ISIS tools are typically needed to achieve high quality networks.

I will modify the wording.

acpaquette commented on isis/src/control/apps/findfeatures/findfeatures.xml:
```diff
@@ -3,431 +3,465 @@
+      the images. These transformations can help improve match quality in
+      specific scenarios. The <b>FASTGEOM</b>, <b>GEOMTYPE</b>, and
+      <b>FASTGEOMPOINTS</b> parameters allow for reprojection of the trainer
+      images into the query image's geometry prior to FBM. (This algorithm is
+      very similar to processing by the ISIS <i>cam2cam</i> application.
+      FASTGEOM differs in that it is not as robust or accurate as
+      <i>cam2cam</i> because FASTGEOM applies a single matrix to warp the
+      entire image - <i>cam2cam</i> uses a rubber sheet projection which is
+      significantly more robust and accurate but can take significantly
+      longer to project images.) These parameters can be used to achieve
```
The information on cam2cam here is decent but seems irrelevant. Largely because nothing in the cam2cam documentation mentions how it operates or that users have any control over that operation.

KrisBecker replied:
For those who understand what cam2cam does, this reference provides context.

Its unfortunate cam2cam documentation does not better explain how it works.

acpaquette commented on isis/src/control/apps/findfeatures/findfeatures.xml:
```diff
@@ -3,431 +3,465 @@
+    <p>
+      The OpenCV methods used in the outlier rejection process have several
+      options that can be set along with the algorithms. The available
+      parameters for are robust marcher algorithms are listed in the following
```
Typo, for are robust marcher -> for the robust matcher algorithms

Not really sure, but that seems to be logically consistent with the rest of the paragraph

acpaquette commented on isis/src/control/apps/findfeatures/findfeatures.xml:
```diff
@@ -730,7 +764,7 @@
           algorithm offers unmatched speed. This speed comes at a cost, the
           BRIEF algorithm behaves very poorly if the query and trainer images
-          vary significantly. The <b>FASTGEOM</b> parameter can help improve
+          vary significantly. The FASTGEOM parameter can help improve
```
Should FASTGEOM be left  bolded here?

KrisBecker replied:
Perhaps. I tried to be consistent. Too much bolding can be distracting from the topic being discussed/emphasized.

I will review.

acpaquette commented on isis/src/control/apps/findfeatures/findfeatures.xml:
```diff
@@ -886,216 +920,816 @@
+    <p>
+      Initial values of outlier matching parameters that can be specified in
+      the PARAMETERS program option are those found in the
+      <a href="#MatcherParameters">Robust Matcher Parameters</a> table.
```
Potential change from "...option are those found in the..." -> "...option can be found in the..."

acpaquette commented on isis/src/control/apps/findfeatures/findfeatures.xml:
```diff
@@ -886,216 +920,816 @@
+      <a href="#MatcherParameters">Robust Matcher Parameters</a> table.
+      These keywords are SaveRenderedImages, SavePath, RootSift,
+      MinimumFundamentalPoints, RefineFundamentalMatrix, and
+      MinimumHomographyPoints. Use of PARAMETERS excludes the need for these
```
If we link to the table why mention the keywords by name here?

acpaquette commented on isis/src/control/apps/findfeatures/findfeatures.xml:
```diff
@@ -886,216 +920,816 @@
+      matrix to each trainer image results in a (fast) projection, or warp,
+      of the image into the image space of the query image, thus minimizing
+      scale and rotation invariance. This defines the FASTGEOM processing
+      objectives in <i>findfeatures</i>. Note, however, that translation,
```
Reads a little easier if ", however," is removed

acpaquette commented on isis/src/control/apps/findfeatures/findfeatures.xml:
```diff
@@ -886,216 +920,816 @@
+      scale and rotation invariance. This defines the FASTGEOM processing
+      objectives in <i>findfeatures</i>. Note, however, that translation,
+      or spatial offsets, may likely exist between images due to the inherent
+      nature of a priori ephemeris data. And, for some image data sets,
+      particularly irregular bodies, this option may not work well,
+      especially for images acquired along the long axis of the target body.
```
The sentence starting with "And," is difficult to understand. Could this change to something like:

"This option is known to have problems with image sets of irregular bodies, especially for images acquired along the long axis of the target body."

KrisBecker replied:
Thanks, I will revise.

acpaquette reviewed (commented):
Hi @KrisBecker, the documentation provided looks good! Couple of things to change here and there. I am not sure about the control network generation tutorial being housed under findfeatures template section as well as the tutorial being placed in the examples of the findfeatures app.

I think a better place to put the tutorial scripts is `isis/scripts/findfeatures_mosaic_tutorial/`, then we can add a new tutorial page under the ISIS wiki in the advanced section called something like "Making Mosaics Using findfeatures"

KrisBecker committed:
Updates to findfeatures PR #4772

- Modified findfeatures.xml documentation to address PR review feedback
- Fixed use of projected images which wasn’t working due to improper instantiation of the cube projection object
- Updated CHANGELOG.md to better categorize all changes in this PR

KrisBecker committed:
Merge remote-tracking branch 'origin/dev' into findfeatures_matrix_fix_mods

KrisBecker commented:
I believe all the comments/concerns have been addressed.

Please let me know if there is anything left to do se we can get this PR merged.

Thanks...

Kelvinrr commented on CHANGELOG.md:
```diff
@@ -35,11 +35,20 @@ release.
 
 ### Added
 - Added rclone to run dependencies in meta.yaml [#5183](https://github.com/DOI-USGS/ISIS3/issues/5183)
+- Add new program optiom <b>TONOGEOM</b> that logs captures geometry errors in the FASTGEOM algorithm and records them to the file provided in this parameter. These images are excluded from the matching process. References [#4772](https://github.com/DOI-USGS/ISIS3/issues/4772)
+- Added a new Radial FASTGEOM transform mapping algorithm to address performance problems with the Grid algorithm. This is now the default algorithm none are selected by the user (see the new <b>GLOBALS</b> parameter to specify the algorithm) [#4772](https://github.com/DOI-USGS/ISIS3/issues/4772)
+- Added new parameter <b>GLOBALS</b> to make parameterization of <i>findfeatures</i> behavior significantly easier and convenient. Wrote significant documentation for the parameter and provide several examples showing its use. [#4772](https://github.com/DOI-USGS/ISIS3/issues/4772)
+- Added two new examples demonstrating/documenting the use of FASTGEOM algorithm, parameterization using <b>GLOBALS</b> and how to produce a regional mosaic using <i>findfeatures</i> with batch scripts. [#4772](https://github.com/DOI-USGS/ISIS3/issues/4772)
+- Added new option <b>GEOMSOURCE=BOTH</b> to check both the MATCH and FROM/FROMLIST images for valid control measure geometry to produce better networks and prevent downstream processing errors. Ignore points that end up with no valid measures (but can retained with use of <b>PreserveIgnoredControl</b> via GLOBALS parameterization). [#4772](https://github.com/DOI-USGS/ISIS3/issues/4772)
+- Added new gtests for <i>findfeatures</i> that replaces all the old application tests. These tests are <i>FunctionalTestFindfeaturesFastGeomDefault</i>, <i>FunctionalTestFindfeaturesFastGeomRadialConfig</i>, <i>FunctionalTestFindfeaturesFastGeomGridDefault</i> and <i>FunctionalTestFindfeaturesFastGeomGridConfig</i>. [#4772](https://github.com/DOI-USGS/ISIS3/issues/4772)
```
As per our discussion at the TC meeting. The fact that his required multiple changelog lines I believe indicates that this should have been split up into multiple PRs.

KrisBecker replied:
I agree this PR became much larger than I expected or even imagined. However, the scope was identified and explained in issue #4772 and in the context of this PR on Feb 3, 2022. At no time did anyone mention this concern even after the initial commits were made.

Clearly, the review process (you are now the third reviewer) has contributed greatly to the size of this PR. Its also clear it did get much larger than anyone expected. I also learned a lot about this process and will be mindful to take steps to prevent this from occurring again. For one thing, I will not let my PRs linger so long between reviews - definitely did not help at all.

I am grateful for the reviews as they have been very helpful and useful to make this PR better. The unfortunate consequence of this is it also made it much larger. Much of the size was in response to requests for documentation improvements/clarifications, which is always useful.

I am open for suggestions/discussion about how this PR could be broken up or made easier and more efficient somehow. Hindsight would indicate to not address two issues in one PR even if you think they are small and/or related.

Kelvinrr replied:
I think at this point, it's too late to break up this PR. It has been through the review process and has had pretty much everything already addressed. But something to consider as a policy to avoid it in the future. Something for the ISIS TC.  

Once the merge conflicts are resolved and the scripts are removed, I'll approve and merge this

Kelvinrr commented on CHANGELOG.md:
```diff
@@ -48,6 +57,10 @@ release.
 ### Fixed
 - Updated History constructor to check for invalid BLOB before copying History BLOB to output cube [#4966](https://github.com/DOI-USGS/ISIS3/issues/4966)
 - Updated photomet MinnaertEmpirical model to support photemplate-style PVL format [#3621](https://github.com/DOI-USGS/ISIS3/issues/3621)
+- Fix matrix inversion errors in findfeatures due to bad FASTGEOM matrix transforms using a more robust implementation to detect these errors and throw exceptions. Images with these errors are captured and logged to the <b>TONOTMATCHED</b> file. Fixes [#4639](https://github.com/DOI-USGS/ISIS3/issues/4639)
+- Fixed use of projected mosaics with correct check for <b>TargetName</b> in the Mapping labels. [#4772](https://github.com/DOI-USGS/ISIS3/issues/4772)
+- Fixed a bug in the <i>cnetwinnow</i> test that did not clean/remove it during test runs.
+- Fixed instantiation and use of projection classes to correctly return geometry data from projected images and mosaics. [#4772](https://github.com/DOI-USGS/ISIS3/issues/4772)
 
```
Even further evidenced that this is a mix of both Bugfixes and feature adds that seem to affect multiple issues.  This would probably make for a good policy discussion.

KrisBecker replied:
The only bug fix in the above entries that was outside the scope of this PR is line 62. It was something I discovered during testing and was fixed in the testing framework.

Kelvinrr commented:
> I believe all the comments/concerns have been addressed.
> 
> Please let me know if there is anything left to do se we can get this PR merged.
> 
> Thanks...

@KrisBecker When I said ping the reviewer, I meant to @ them so they get pinged. Adam has been busy putting out some fires so I'm gonna take over review work with his permission. I agree it's good to go except for the scripts which is the odd man out.

Kelvinrr reviewed (changes requested):
I think this is good except we should remove those tutorial scripts. Or someone needs to make a convincing argument that it's fine. 

They can be reintegrated later down the line. My desire to see this merged is only slightly lower than my desire to avoid introducing more tech debt.

KrisBecker committed:
Removed scripts in example 4 of findfeatures docs

- Per request via USGS review, removed the Bash shell scripts that produce the results of example 4

- Removed the $ISISROOT/appdata/templates/findfeatures/mosaics containing the scripts

- Updated documentation in findfeatures.xml accordingly

- Updated CHANGELOG.md accordingly

KrisBecker committed:
Merge remote-tracking branch 'origin/dev' into findfeatures_matrix_fix_mods

# Conflicts:
#	CHANGELOG.md

KrisBecker committed:
Merge branch 'DOI-USGS:dev' into findfeatures_matrix_fix_mods

KrisBecker commented:
@Kelvinrr updates have been completed and conflict resolved.

Let me know if there is anything remaining.

Thanks...
